pkg/web/context_keys.go

```diff
@@ -32,3 +35,25 @@ func IsAuthorized(ctx context.Context) bool {
 func ContextWithAuthorization(ctx context.Context) context.Context {
 	return context.WithValue(ctx, isAuthorizedKey, true)
 }
+
+func ContextWithAuthenticationError(ctx context.Context, authNError error) context.Context {
+	return context.WithValue(ctx, authenticationErrorKey, authNError)
+}
+
+func AuthenticationErrorFromContext(ctx context.Context) (bool, error) {
+	authnError, ok := ctx.Value(authenticationErrorKey).(error)
+	return ok && authnError != nil, authnError
+}
+
+func ContextWithAuthorizationError(ctx context.Context, authZError error) context.Context {
+	found, currentAuthZError := AuthorizationErrorFromContext(ctx)
+	if found {
+		authZError = fmt.Errorf("%s or %s", currentAuthZError, authZError)
```

Comment:
It is a composition of errors, which are independent, so it is OR of errors. It can happen that there is AND in the error if there are checks which are all required